complex objects/cargo/list/src/lib.rs

```diff
@@ -157,74 +172,77 @@ impl ListManager {
-                None
+        let mut item_iter = stmt.query_map(&[uuid], |row| {
+            Item {
+                uuid: row.get(0),
+                name: row.get(1),
+                due_date: row.get(2),
+                completion_date: row.get(3),
+                labels: self.fetch_labels_for_item(uuid)
             }
+        }).unwrap();
+
+        if let Some(result) = item_iter.next() {
+            result.ok()
+        } else {
+            println!("No item found for uuid {:?}", uuid);
+            None
         }
     }
 
+    pub fn create_item(&self, item: &Item) -> Option<Item> {
+        println!("Creating item {:?}", item);
+        let item_sql = r#"INSERT INTO items (uuid, name, due_date, completion_date) VALUES (?, ?, ?, ?)"#;
+        let conn = self.store.write_connection();
+        let item_uuid = Uuid::new_v4().simple().to_string();
+        conn.execute(item_sql, &[&item_uuid, &item.name, &item.due_date, &item.completion_date]).unwrap();
+
+        let item_label_sql = r#"INSERT INTO item_labels (item_uuid, label_name) VALUES (?, ?)"#;
+        let _ = item.labels.iter().map(|label| {
+            conn.execute(&item_label_sql, &[&item_uuid, &label.name]).unwrap();
+        });
+        self.fetch_item(&item_uuid)
+    }
+
     pub fn update_item(&self, item: &Item) {
-        let sql = r#"UPDATE items SET description=?, due_date=?, is_complete=? WHERE id=?"#;
+        let sql = r#"UPDATE items SET name=?, due_date=?, completion_date=? WHERE uuid=?"#;
         let conn = self.store.write_connection();
-        let _ = conn.execute(sql, &[&item.description, &item.due_date, &item.is_complete, &item.id]);
+        let _ = conn.execute(sql, &[&item.name, &item.due_date, &item.completion_date, &item.uuid]);
     }
 }
 
 
 #[no_mangle]
-pub unsafe extern "C" fn get_all_categories(manager: *const Arc<ListManager>) -> *mut Vec<Category> {
+pub unsafe extern "C" fn list_manager_get_all_labels(manager: *const Arc<ListManager>) -> *mut Vec<Label> {
     let manager = &*manager;
-    let category_list = Box::new(manager.fetch_categories());
-    Box::into_raw(category_list)
+    let label_list = Box::new(manager.fetch_labels());
+    Box::into_raw(label_list)
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn category_manager_create_item(manager: *const Arc<ListManager>, item: *const Item, category_id: c_int) {
+pub unsafe extern "C" fn list_manager_create_item(manager: *const Arc<ListManager>, item: *const Item) {
     let manager = &*manager;
     let item = &*item;
-    let cat_id = category_id as i64;
-    manager.create_item(item, cat_id);
+    manager.create_item(item);
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn category_manager_update_item(manager: *const Arc<ListManager>, item: *const Item) {
+pub unsafe extern "C" fn list_manager_update_item(manager: *const Arc<ListManager>, item: *const Item) {
     let manager = &*manager;
     let item = &*item;
     manager.update_item(item)
```

Comment:
Ah, so to update an entitiy client has to perform two api calls. "set {field} on {entity}" followed by "update {entity}".

Reply:
Yeah, we should probably change the FFI API to better reflect the one we described for the WebExtension API, but I will do that in another PR. This one is just about changing the data model.